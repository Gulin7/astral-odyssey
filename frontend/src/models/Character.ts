class Character {
    private id: number;
    private name: string;
    private charClass: string;
    private race: string;
    private playerId: number;
    private skinUrl: string;
    private level: number;

    public constructor(
        id: number,
        name: string,
        charClass: string,
        race: string,
        playerId: number,
    ) {
        this.id = id;
        this.name = name;
        this.charClass = charClass;
        this.race = race;
        this.playerId = playerId;
        this.skinUrl = this.getDefaultSkinUrl();
        this.level = 1;
    }

    public getDefaultSkinUrl(): string {
        if (this.charClass === 'Warrior') {
            return 'warrior-default.png';
        } else if (this.charClass === 'Mage') {
            return 'mage-default.png';
        } else if (this.charClass === 'Fighter') {
            return 'fighter-default.png';
        } else if (this.charClass === 'Ranger') {
            return 'ranger-default.png';
        }
        return 'warrior-default';
    }

    public getId(): number {
        return this.id;
    }

    public setId(newId: number): void {
        this.id = newId;
    }

    public getName(): string {
        return this.name;
    }

    public setName(newName: string): void {
        this.name = newName;
    }

    public getCharClass(): string {
        return this.charClass;
    }

    public setCharClass(newClass: string): void {
        this.charClass = newClass;
        this.setSkinUrl(this.getDefaultSkinUrl());
    }

    public getRace(): string {
        return this.race;
    }

    public setRace(newRace: string): void {
        this.race = newRace;
    }

    public getPlayerId(): number {
        return this.playerId;
    }

    public setPlayerId(newPlayerId: number): void {
        this.playerId = newPlayerId;
    }

    public getSkinUrl(): string {
        return this.skinUrl;
    }

    public setSkinUrl(newUrl: string): void {
        this.skinUrl = newUrl;
    }

    public getLevel(): number {
        return this.level;
    }

    public setLevel(newLevel: number): void {
        this.level = newLevel;
    }
}

export default Character;
