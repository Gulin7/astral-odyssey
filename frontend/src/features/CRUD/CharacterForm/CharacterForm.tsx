import Character from '../../../../../models/Character';
import {CharacterFormType} from '../../../utils/types/CharacterFormProps.types';

import {FormEntryProps} from '../../../utils/types/FormEntryProps.types';
import {CharacterFormEntry} from '../CharacterFormEntry/CharacterFormEntry';

import './CharacterForm.css';

function setCharacterFormEntriesForCharacter(
    formEntries: FormEntryProps[],
    givenCharacter: Character | undefined,
) {
    if (givenCharacter !== undefined) {
        formEntries[0].disabled = true;

        formEntries[0].defaultValue = givenCharacter.getId().toString();
        formEntries[1].defaultValue = givenCharacter.getName();
        formEntries[2].defaultValue = givenCharacter.getCharClass();
        formEntries[3].defaultValue = givenCharacter.getRace();
        formEntries[4].defaultValue = givenCharacter.getPlayerId().toString();
    }

    return formEntries;
}

function createFormEntries(props: CharacterFormType) {
    let formEntries = [
        {
            label: 'Character ID',
            ref: props.idInput,
            placeHolder: 'Enter ID',
            defaultValue: '',
            disabled: false,
        },
        {
            label: 'Character Name',
            ref: props.nameInput,
            placeHolder: 'Enter name',
            defaultValue: '',
            disabled: false,
        },
        {
            label: 'Character Class',
            ref: props.classInput,
            placeHolder: 'Enter class',
            defaultValue: '',
            disabled: false,
        },
        {
            label: 'Character Race',
            ref: props.raceInput,
            placeHolder: 'Enter race',
            defaultValue: '',
            disabled: false,
        },
        {
            label: 'Player ID',
            ref: props.playerIdInput,
            placeHolder: 'Enter player ID',
            defaultValue: '',
            disabled: false,
        },
    ];

    formEntries = setCharacterFormEntriesForCharacter(
        formEntries,
        props.givenCharacter,
    );

    return formEntries;
}

export function CharacterForm(props: CharacterFormType) {
    const formEntries = createFormEntries(props);

    return (
        <div
            className='character-form-div'
            data-testid='character-form-test-id'
        >
            <form className='character-form'>
                {formEntries.map((entry) => (
                    <CharacterFormEntry
                        key={entry.label}
                        ref={entry.ref}
                        label={entry.label}
                        placeHolder={entry.placeHolder}
                        defaultValue={entry.defaultValue}
                        disabled={entry.disabled}
                    />
                ))}
            </form>
        </div>
    );
}
