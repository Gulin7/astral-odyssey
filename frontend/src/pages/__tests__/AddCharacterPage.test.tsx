import '@testing-library/jest-dom';
import {render, screen} from '@testing-library/react';
import {expect, test} from 'vitest';

import React from 'react';
import Character from '../../../../models/Character';
import {CharacterForm} from '../../features/CRUD/CharacterForm/CharacterForm';

test('testing rendering of character form without character', () => {
    let idInput = React.createRef<HTMLInputElement>();
    let nameInput = React.createRef<HTMLInputElement>();
    let classInput = React.createRef<HTMLInputElement>();
    let raceInput = React.createRef<HTMLInputElement>();
    let playerIdInput = React.createRef<HTMLInputElement>();

    const renderedComp = render(
        <CharacterForm
            idInput={idInput}
            nameInput={nameInput}
            classInput={classInput}
            raceInput={raceInput}
            playerIdInput={playerIdInput}
        />,
    );

    const renderedCharacterForm = screen.getByTestId('character-form-test-id');
    const idFormInput = screen.getByPlaceholderText('Enter ID');
    const nameFormInput = screen.getByPlaceholderText('Enter name');
    const classFormInput = screen.getByPlaceholderText('Enter class');
    const raceFormInput = screen.getByPlaceholderText('Enter race');

    const idFormLabel = screen.getByText('Character ID');
    const nameFormLabel = screen.getByText('Character Name');
    const classFormLabel = screen.getByText('Character Class');

    expect(renderedCharacterForm).toBeInTheDocument();
    expect(idFormInput).toBeInTheDocument();
    expect(nameFormInput).toBeInTheDocument();
    expect(classFormInput).toBeInTheDocument();
    expect(raceFormInput).toBeInTheDocument();
    expect(idFormLabel).toBeInTheDocument();
    expect(nameFormLabel).toBeInTheDocument();
    expect(classFormLabel).toBeInTheDocument();
    expect(renderedComp).toMatchSnapshot();
});

test('testing rendering of character form with character', () => {
    let idInput = React.createRef<HTMLInputElement>();
    let nameInput = React.createRef<HTMLInputElement>();
    let classInput = React.createRef<HTMLInputElement>();
    let raceInput = React.createRef<HTMLInputElement>();
    let playerIdInput = React.createRef<HTMLInputElement>();

    let demoCharacter = new Character(
        1,
        'testName',
        'testClass',
        'testRace',
        2,
    );

    render(
        <CharacterForm
            idInput={idInput}
            nameInput={nameInput}
            classInput={classInput}
            raceInput={raceInput}
            playerIdInput={playerIdInput}
            givenCharacter={demoCharacter}
        />,
    );

    const renderedPlayerForm = screen.getByTestId('character-form-test-id');
    const idFormInput = screen.getByDisplayValue('1');
    const nameInputFormInput = screen.getByDisplayValue('testName');
    const classInputFormInput = screen.getByDisplayValue('testClass');
    const raceInputFormInput = screen.getByDisplayValue('testRace');
    const playerIdInputFormInput = screen.getByDisplayValue('2');

    expect(renderedPlayerForm).toBeInTheDocument();
    expect(idFormInput).toBeInTheDocument();
    expect(nameInputFormInput).toBeInTheDocument();
    expect(classInputFormInput).toBeInTheDocument();
    expect(raceInputFormInput).toBeInTheDocument();
    expect(playerIdInputFormInput).toBeInTheDocument();
});
