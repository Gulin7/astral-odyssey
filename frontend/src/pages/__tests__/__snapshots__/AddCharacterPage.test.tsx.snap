// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`testing rendering of character form without character 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="character-form-div"
        data-testid="character-form-test-id"
      >
        <form
          class="character-form"
        >
          <div
            class="character-form-entry"
            data-testid="character-form-entry-test-id"
          >
            <label
              class="character-form-label"
              for="Character ID"
            >
              Character ID
            </label>
            <input
              class="character-form-input"
              data-testid="character-form-input-test-id"
              id="Character ID"
              placeholder="Enter ID"
              type="text"
            />
          </div>
          <div
            class="character-form-entry"
            data-testid="character-form-entry-test-id"
          >
            <label
              class="character-form-label"
              for="Character Name"
            >
              Character Name
            </label>
            <input
              class="character-form-input"
              data-testid="character-form-input-test-id"
              id="Character Name"
              placeholder="Enter name"
              type="text"
            />
          </div>
          <div
            class="character-form-entry"
            data-testid="character-form-entry-test-id"
          >
            <label
              class="character-form-label"
              for="Character Class"
            >
              Character Class
            </label>
            <input
              class="character-form-input"
              data-testid="character-form-input-test-id"
              id="Character Class"
              placeholder="Enter class"
              type="text"
            />
          </div>
          <div
            class="character-form-entry"
            data-testid="character-form-entry-test-id"
          >
            <label
              class="character-form-label"
              for="Character Race"
            >
              Character Race
            </label>
            <input
              class="character-form-input"
              data-testid="character-form-input-test-id"
              id="Character Race"
              placeholder="Enter race"
              type="text"
            />
          </div>
          <div
            class="character-form-entry"
            data-testid="character-form-entry-test-id"
          >
            <label
              class="character-form-label"
              for="Player ID"
            >
              Player ID
            </label>
            <input
              class="character-form-input"
              data-testid="character-form-input-test-id"
              id="Player ID"
              placeholder="Enter player ID"
              type="text"
            />
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="character-form-div"
      data-testid="character-form-test-id"
    >
      <form
        class="character-form"
      >
        <div
          class="character-form-entry"
          data-testid="character-form-entry-test-id"
        >
          <label
            class="character-form-label"
            for="Character ID"
          >
            Character ID
          </label>
          <input
            class="character-form-input"
            data-testid="character-form-input-test-id"
            id="Character ID"
            placeholder="Enter ID"
            type="text"
          />
        </div>
        <div
          class="character-form-entry"
          data-testid="character-form-entry-test-id"
        >
          <label
            class="character-form-label"
            for="Character Name"
          >
            Character Name
          </label>
          <input
            class="character-form-input"
            data-testid="character-form-input-test-id"
            id="Character Name"
            placeholder="Enter name"
            type="text"
          />
        </div>
        <div
          class="character-form-entry"
          data-testid="character-form-entry-test-id"
        >
          <label
            class="character-form-label"
            for="Character Class"
          >
            Character Class
          </label>
          <input
            class="character-form-input"
            data-testid="character-form-input-test-id"
            id="Character Class"
            placeholder="Enter class"
            type="text"
          />
        </div>
        <div
          class="character-form-entry"
          data-testid="character-form-entry-test-id"
        >
          <label
            class="character-form-label"
            for="Character Race"
          >
            Character Race
          </label>
          <input
            class="character-form-input"
            data-testid="character-form-input-test-id"
            id="Character Race"
            placeholder="Enter race"
            type="text"
          />
        </div>
        <div
          class="character-form-entry"
          data-testid="character-form-entry-test-id"
        >
          <label
            class="character-form-label"
            for="Player ID"
          >
            Player ID
          </label>
          <input
            class="character-form-input"
            data-testid="character-form-input-test-id"
            id="Player ID"
            placeholder="Enter player ID"
            type="text"
          />
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
